### PR TITLE
Track websocket consumers, expose them via an admin endpoint, list th…

### PR DESCRIPTION
**Pull request by ericvolp12**
…em in the UI, clean up connections when they stop responding

CleanShot 2023-07-20 at 10 58 20

### CODE REVIEW

File: bgs/bgs.go [hunk 1]
```diff
@@ -394,26 +413,86 @@ func (bgs *BGS) EventsHandler(c echo.Context) error {
...
+	defer cleanup()
+
+	// Keep track of the consumer for metrics and admin endpoints
+	consumer := SocketConsumer{
+		RemoteAddr:  c.RealIP(),
+		UserAgent:   c.Request().UserAgent(),
+		ConnectedAt: time.Now(),
+	}
+
+	sentCounter := eventsSentCounter.WithLabelValues(consumer.RemoteAddr, consumer.UserAgent)
```
Comment (whyrusleeping) [inline bgs/bgs.go:459]: the cardinality on this metric is probably going to be very high

Comment (ericvolp12) [reply bgs/bgs.go:459]: I'm guessing it'll be under 10k series initially, basically one series per unique consumer (though repeat consumers will end up using the same series). If it starts to get too high we can try reducing it but I like the idea of remote address and/or user-agent tracking in a visible way.

Comment (ericvolp12) [reply bgs/bgs.go:459]: Also it'll get GC'd on BGS restart so we don't have a ton of inactive series to track (hopefully).

Comment (ericvolp12) [reply bgs/bgs.go:459]: We could maybe track it internally at this level and via prometheus only at the RemoteAddr level to start

File: bgs/bgs.go [hunk 2]
```diff
@@ -394,26 +413,86 @@ func (bgs *BGS) EventsHandler(c echo.Context) error {
...
+	consumer.EventsSent = sentCounter
+
+	bgs.consumersLk.Lock()
```
Comment (whyrusleeping) [inline bgs/bgs.go:462]: probably deserves a `bgs.registerConsumer()` method

File: bgs/bgs.go [hunk 3]
```diff
@@ -394,26 +413,86 @@ func (bgs *BGS) EventsHandler(c echo.Context) error {
...
+	bgs.nextConsumerID++
+	consumerID := bgs.nextConsumerID
+	bgs.consumers[bgs.nextConsumerID] = &consumer
+	bgs.consumersLk.Unlock()
+
+	// Cleanup the consumer when we're done
+	defer func() {
```
Comment (whyrusleeping) [inline bgs/bgs.go:469]: and `bgs.cleanupConsumer()`

File: ts/bgs-dash/src/components/Consumers/Consumers.tsx [hunk 4]
```diff
@@ -0,0 +1,352 @@
+import { ChevronDownIcon, ChevronUpIcon } from "@heroicons/react/24/solid";
```
Comment (whyrusleeping) [inline ts/bgs-dash/src/components/Consumers/Consumers.tsx:1]: image

File: bgs/metrics.go [hunk 5]
```diff
@@ -19,3 +19,8 @@ var rebasesCounter = promauto.NewCounterVec(prometheus.CounterOpts{
 	Name: "event_rebases",
 	Help: "The total number of rebase events received",
 }, []string{"pds"})
+
+var eventsSentCounter = promauto.NewCounterVec(prometheus.CounterOpts{
+	Name: "events_sent_counter",
+	Help: "The total number of events sent to consumers",
+}, []string{"remote_addr", "user_agent"})
```
Comment (whyrusleeping) [inline bgs/metrics.go:26]: smells like lots of metrics

File: bgs/bgs.go (hunk 1 above)
Comment (whyrusleeping) [reply bgs/bgs.go:490]: yeah i guess that seems fine for now, lets just keep an eye on it

**Review by whyrusleeping — APPROVED** [review]
LGTM

**Pull request merged**